### ISSUE TITLE
Simple way to run benchmark with a ref of a state

### ISSUE THREAD

**Issue by jpochyla** [bazhenov/tango#16]
I understand it's possible to use the matrix + generators for this, but something akin to this simple Criterion code is currently missing, I think:

```
fn benchmark(c: &mut Criterion) {
    let mut state = compute_heavy_state();
    c.bench_function("foo", |b| {
        b.iter(|| bar(&mut state))
    });
}
```

**Comment by bazhenov** [comment]
Now as I'm thinking about it, it seems easy to implement function similar to `benchmark_fn()`, that will pass state by reference:

```rust
fn bench() -> imp IntoBenchmark {
  benchmark_fn_state(compute_heavy_state(), |state| {
    // do work
  })
}
```

**Comment by FilipAndersson245** [comment]
Any update on this issue?
I have a large mutatable state that I would rather clone and insert to the benchmark, rather than calculate it during benchmarking.

**Comment by bazhenov** [comment]
Yes, there is new API in `dev` branch that is very similar to criterion API that is solves the issue. README.md in `dev` branch has some simple examples. Please let me know if you have any issues with it.

**Comment by FilipAndersson245** [comment]
I was looking over the code on dev and as I understand its working it would still require me to clone inside the benchmarking function cloneing the eventual heavy state, I think how you can do it in a separate setup function in Criterion is decently clean.

```rs
fn bench_load_standard_lib(c: &mut Criterion) {
    c.bench_function("load_standard_lib", |b| {
        let engine = setup_engine();
        b.iter_batched(
            || engine.clone(),
            |mut engine| {
                load_standard_library(&mut engine).unwrap();
            },
            criterion::BatchSize::SmallInput,
        );
    });
}
```
Doing the equivalent here would still require the cloning to be done inside the `b.iter()` function?

**Comment by bazhenov** [comment]
You don't have to clone the state. You can clone the `Rc` ref. There is a relevant example in source code:

https://github.com/bazhenov/tango/blob/51f4e8414e4aabefda987baa2f26fa8a7c72110f/examples/benches/test_funcs.rs#L9-L20

The reason behind it that tango can rotate the state at a given amount of samples (see: `MeasurementSettings::samples_per_haystack`). This way you can control effects of CPU-caching behavior.

**Comment by FilipAndersson245** [comment]
But if I need a mutatable state i cannot use that right?

**Comment by bazhenov** [comment]
Yes, mutable state is not allowed at the moment. This makes benchmarks less reproducible, because each iteration is not independent in case of mutable state.

**Comment by FilipAndersson245** [comment]
Currently when testing out I need to do something like this to get access to the state, that has to be reset between iterations.

```rs
fn load_standard_lib() -> impl IntoBenchmarks {
    let engine = setup_engine();
    [benchmark_fn("load_standard_lib", move |b| {
        let engine = engine.clone();
        b.iter(move || {
            let mut engine = engine.clone();
            load_standard_library(&mut engine)
        })
    })]
}```

**Comment by bazhenov** [comment]
Just to make sure I understand you correctly. You want some nice API to be able to clone mutable state when defining benchmark function. Something along the lines of

```rust
fn load_standard_lib() -> impl IntoBenchmarks {
    let engine = setup_engine();
    [benchmark_fn("load_standard_lib", move |b| {
        b.iter_with_state(|| engine.clone(), |e| {
          load_standard_library(&mut e)
        })
    })]
}
```

**Comment by FilipAndersson245** [comment]
Yes, that would be very simular to how criterion does, as we have a large mut state and prefer not to have the cloning be part of the measurment.
